swiftide-integrations/src/pgvector/pgv_table_types.rs

```diff
@@ -0,0 +1,481 @@
+                            .or_default();
+
+                        let mut metadata_map = BTreeMap::new();
+                        metadata_map.insert(config.original_field.clone(), value.clone());
+                        entry.push(serde_json::to_value(metadata_map)?);
+                    }
+                    FieldConfig::Vector(config) => {
+                        let data = node
+                            .vectors
+                            .as_ref()
+                            .and_then(|v| v.get(&config.embedded_field))
+                            .map(|v| v.to_vec())
+                            .unwrap_or_default();
+
+                        bulk_data
+                            .vector_fields
+                            .entry(config.field.clone())
+                            .or_default()
+                            .push(ExtPgVector::Vector::from(data));
+                    }
+                    _ => continue, // ID and Chunk already handled
+                }
+            }
+        }
+
+        Ok(bulk_data)
+    }
+
+    /// Generates SQL for UNNEST-based bulk upsert.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<String>` - The generated SQL statement or an error if fields are empty.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if `self.fields` is empty, as no valid SQL can be generated.
+    fn generate_unnest_upsert_sql(&self) -> Result<String> {
+        if self.fields.is_empty() {
+            return Err(anyhow!("Cannot generate upsert SQL with empty fields"));
+        }
+
+        let mut columns = Vec::new();
+        let mut unnest_params = Vec::new();
+        let mut param_counter = 1;
+
+        for field in &self.fields {
+            let name = field.field_name();
+            columns.push(name.to_string());
+
+            unnest_params.push(format!(
+                "${param_counter}::{}",
+                match field {
+                    FieldConfig::ID => "UUID[]",
+                    FieldConfig::Chunk => "TEXT[]",
+                    FieldConfig::Metadata(_) => "JSONB[]",
+                    FieldConfig::Vector(_) => "VECTOR[]",
+                }
+            ));
+
+            param_counter += 1;
+        }
+
+        let update_columns = self
+            .fields
+            .iter()
+            .filter(|field| !matches!(field, FieldConfig::ID)) // Skip ID field in updates
+            .map(|field| {
+                let name = field.field_name();
+                format!("{name} = EXCLUDED.{name}")
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        Ok(format!(
+            r#"
+            INSERT INTO {} ({})
+            SELECT {}
+            FROM UNNEST({}) AS t({})
+            ON CONFLICT (id) DO UPDATE SET {}"#,
+            self.table_name,
+            columns.join(", "),
+            columns.join(", "),
+            unnest_params.join(", "),
+            columns.join(", "),
+            update_columns
```

Comment:
Hi @timonv,

I apologize for the delay.

- Regarding your comment "Additionally, it's generated for each bulk insert, but I think it's identical for the whole pipeline, right?"

Yes you were right that the SQL statement for bulk insert is applied to the entire pipeline. Optimizing this could be achieved by introducing a new state variable in `PgVector` and initializing it as part of `setup()`, does it makesense? :

```rust
#[derive(Builder, Clone)]
#[builder(setter(into, strip_option), build_fn(error = "anyhow::Error"))]
pub struct PgVector {
    sql_stmt_bulk_insert: String,
}
```

- For your comment "If columns or table names include sql, this is vulnerable to sql injection. There has to be a way to avoid this?", I need your help and more details. Please find the SQL statement used for bulk insert:

```sql
INSERT INTO swiftide_pgvector_test (
    id,
    chunk,
    vector_combined,
    meta_questions_and_answers
)
SELECT 
    id,
    chunk,
    vector_combined,
    meta_questions_and_answers
FROM UNNEST(
    $1::UUID[],
    $2::TEXT[],
    $3::VECTOR[],
    $4::JSONB[]
) AS t(
    id,
    chunk,
    vector_combined,
    meta_questions_and_answers
)
ON CONFLICT (id) DO UPDATE 
SET 
    chunk = EXCLUDED.chunk,
    vector_combined = EXCLUDED.vector_combined,
    meta_questions_and_answers = EXCLUDED.meta_questions_and_answers;
```

Could you please clarify what might be causing the SQL injection vulnerability? Are there any issues with the naming conventions used for `table` and `column_name`, or is it related to the usage of placeholders?

Thank you for your help.

Reply:
As for optimizing the bulk insert, a local static perhaps. For the sql injection I suspect the column names might be vulnerable as they're generated at runtime without escaping, but the combination of it producing a working query and it actually happening (table names from user/ai input?) are minor. I'm fine with leaving it as is, can always be picked up later.